design/client-side-encryption_design.md

```diff
@@ -0,0 +1,461 @@
+    if err != nil {
+        ctx.log.Errorf("error decrypting backup: %s", err.Error())
+        errs.AddVeleroError(err)
+        return warnings, errs
+    }
+    
+    } else {
+        backupContent = ctx.backupReader
+    }
+    
+    dir, err := archive.NewExtractor(ctx.log, ctx.fileSystem).UnzipAndExtractBackup(backupContent)
+	...
+}
+```
+
+## Alternatives Considered
+
+This could also be implemented in object store plugins.
+Then however, it has to be implemented in every single object store plugin.
+This is not viable.
+
+We could just make use of server-side encryption of an object store.
+But what about object stores that do not support this?
+Also, client-side encryption is a lot safer, since the backups are encrypted during transit as well.
+
+## Security Considerations
+
+Overall this feature should increase the security of the product by encrypting secrets and application configurations.
+However, the security of this feature has to be ensured to avoid giving users a false sense of security.
+Since only the main backup archive is encrypted, it should be documented precisely what information gets encrypted and what not.
+Unencrypted information includes:
+- CSI-VolumeSnapshotClasses
+- CSI-VolumeSnapshotContents (despite the name, these do not contain the contents of the backup)
+- CSI-VolumeSnapshots
+- Item-Operations
+- Backup-Logs
+- PodVolumeBackups
+- a resource list containing the names and kinds of the backed up resources
+- Results (errors, warnings) of the backup operation
+- (native) VolumeSnapshots
+- the velero backup custom resource
+
+If the secrets containing the encryption keys are deleted, users might lose access to their backups.
+Therefore, users should back up their encryption keys outside the cluster.
+
+## Compatibility
+
+Encrypting the whole backup would introduce major breaking changes for object store plugins and maybe even more.
+Therefore, only the main backup archive is encrypted.
+
+If unencrypted backups exist when the feature is activated, they will stay unencrypted and can still be restored.
+Likewise, when the feature gets disabled, encrypted backups can still be decrypted.
+It is also possible to change the encryption key without loosing access to previous backups.
+This is due to the backups containing metadata about their encryption.
+If this metadata is absent, the backup is assumed to be unencrypted.
+Currently, this encryption status contains a flag whether the backup is encrypted, as well as the name of the secret containing the key it was encrypted with.
+In the future, the encryption status can be extended by the type of encryption without loosing backwards compatability by assuming the default to be AES.
+
+There should be no breaking changes.
+
+## Implementation
+
+A working prototype already exists.
+Therefore, minimal effort is needed to implement this.
+Timelines are almost entirely dependent on how fast this goes through and how many things need changing before it can.
```

Comment:
This sentence seems to be unfinished.
```suggestion
Timelines are almost entirely dependent on how fast this goes through and how many things need to be changed before it can be merged.
```